Snap 1334 : Auto Refresh feature for Dashboard UI

snappy-sachin opened the pull request:
## Changes proposed in this pull request

 - Adding ClusterStatistics and MemberStatistics classes for maintaining Cluster level and Member level Statistics.
 - Updated gradle build fie for org.apache.commons:commons-collections4:4.1

## Patch testing

 - Manually

## ReleaseNotes changes

 - NA

## Other PRs 

 - SnappyData: https://github.com/SnappyDataInc/snappydata/pull/1005
 - Spark: https://github.com/SnappyDataInc/spark/pull/99

snappy-sachin committed:
Changes for SNAP-1334:
 - Updating dependencies in build.gradle for adding apache commons collection4 org.apache.commons:commons-collections4:4.1
 - Updating MemberStatisticsMessage to include collection time of members stats.
 - Adding MemberStatistics class which holds members stats details.

snappy-sachin committed:
Merge branch 'snappy/master' into SNAP-1334

snappy-sachin committed:
Code changes for SNAP-1334:
- Adding Singleton class ClusterStatistics to hold Cluster level stats and usage trends.

snappy-sachin committed:
Merge branch 'snappy/master' into SNAP-1334

snappy-sachin committed:
Code changes for SNAP-1334:
 - Changes for agrregate memory (heap + off-heap) usage trends.

snappy-sachin committed:
Merge branch 'snappy/master' into SNAP-1334

snappy-sachin committed:
Merge branch 'snappy/master' into SNAP-1334

snappy-sachin committed:
Code changes for SNAP-1334:
 - Displaying Heap, Off-Heap and Memory Usage line charts in GB instead of in percentages.
 - Adding SnappyUtils.java Utility class.

snappy-sachin committed:
Merge branch 'snappy/master' into SNAP-1334

snappy-sachin committed:
Merge branch 'snappy/master' into SNAP-1334

snappy-sachin committed:
Code changes for SNAP-1334:
  - Adding disk store disk space usage details in members and clusters statistics.

snappy-sachin committed:
Code changes for SNAP-1334:
 - Consider host only once if multiple nodes are running on same host.
 - Also do not consider Locator node while calculating cluster level CPU usage.

snappy-sachin committed:
Merge branch 'snappy/master' into SNAP-1334

rishitesh commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ui/ClusterStatistics.java:
```diff
@@ -0,0 +1,239 @@
+        }
+        break;
+      case TREND_HEAP_USAGE:
+        synchronized (this.heapUsageTrend) {
```
Why synchronize if we are not synchronizing while adding data to queue? We should think about a concurrent structure which is bounded &  non-blocking ( If this method will be executed every time for each browser request. Lets say if multiple users are accessing the dashboard how many times this method gets called ? )
If not simply adding a synchronized keyword while adding to the queue will suffice.

snappy-sachin replied:
As suggested, removed synchronized keyword where present while reading values from queue. Now Synchronizing only when adding values to queue.

rishitesh commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ui/SnappyUtils.java:
```diff
@@ -0,0 +1,94 @@
+      System.out.println("in KB");
+      value = (double) size / KB;
+    } else {
+      System.out.println("in ELSE");
```
remove SOP statements.

snappy-sachin replied:
Done.

rishitesh commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ui/SnappyUtils.java:
```diff
@@ -0,0 +1,94 @@
+   * Convert a quantity in bytes to a human-readable string in specified units.
+   */
+  public static double bytesToGivenUnits(long size, StorageSizeUnits expectedUnit) {
+    long TB = 1L << 40;
```
can be static final values.

snappy-sachin replied:
As suggested, replaced enum by static final string constants.

rishitesh commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ui/SnappyUtils.java:
```diff
@@ -0,0 +1,94 @@
+public class SnappyUtils {
+
+  public enum StorageSizeUnits {
+    B, KB, MB, GB, TB, ANY
```
Should add support for PB also. If we scale in some cases this should not be a re-work.

snappy-sachin replied:
Added support for PB (peta bytes).

rishitesh commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ui/SnappyUtils.java:
```diff
@@ -0,0 +1,94 @@
+      unit = "B";
+    }
+
+    return String.format(Locale.US, "%.1f %s", value, unit);
```
Should use Locale.default() rather than US.

snappy-sachin replied:
Done. using Locale.getDefault()

snappy-sachin committed:
Merge branch 'snappy/master' into SNAP-1334

snappy-sachin committed:
Code changes for SNAP-1334:
 - SOP statements clean up.
 - Adding support for PB (peta bytes) size.
 - String locale changed from US to Default.

snappy-sachin committed:
Code changes for SNAP-1334:
 - Using static final instead of enum for defining storage/memory size units.

snappy-sachin committed:
Code changes for SNAP-1334:
 - Adding a synchronized keyword while adding data to the queue.
 - Removing a synchronized keyword where just reading data from the queue.

snappy-sachin commented:
All feedback comments have been taken care.

snappy-sachin committed:
Code changes for SNAP-1334:
 - Unused variable clean up.
